apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/functions/AllocateAndAddFloatingIpToNode.java

```diff
@@ -78,37 +94,77 @@ public AllocateAndAddFloatingIpToNode(@Named(TIMEOUT_NODE_RUNNING) Predicate<Ato
    @Override
    public AtomicReference<NodeMetadata> apply(AtomicReference<NodeAndNovaTemplateOptions> input) {
       checkState(nodeRunning.apply(input.get().getNodeMetadata()), "node never achieved state running %s", input.get().getNodeMetadata());
-      NodeMetadata node = input.get().getNodeMetadata().get();
+      final NodeMetadata node = input.get().getNodeMetadata().get();
       // node's location is a host
       String regionId = node.getLocation().getParent().getId();
-      FloatingIPApi floatingIpApi = novaApi.getFloatingIPApi(regionId).get();
       Optional<Set<String>> poolNames = input.get().getNovaTemplateOptions().get().getFloatingIpPoolNames();
 
-      Optional<FloatingIP> ip = allocateFloatingIPForNode(floatingIpApi, poolNames, node.getId());
-      if (!ip.isPresent()) {
-         cleanupResources.apply(node);
-         throw new InsufficientResourcesException("Failed to allocate a FloatingIP for node(" + node.getId() + ")");
-      }
-      logger.debug(">> adding floatingIp(%s) to node(%s)", ip.get().getIp(), node.getId());
+      String availabilityZone = getAvailabilityZoneFromTemplateOptionsOrDefault(input, regionId);
+
+      if (isNeutronLinked()) {
+         org.jclouds.openstack.neutron.v2.features.FloatingIPApi neutronFloatingApi = getFloatingIPApi(regionId);
+            final Optional<Port> optionalPort = getPortApi(regionId).list().concat().firstMatch(new Predicate<Port>() {
+               @Override
+               public boolean apply(@Nullable Port input) {
+                  return input.getDeviceId().equals(node.getProviderId());
+               }
+            });
+            if (optionalPort.isPresent()) {
+               Optional<org.jclouds.openstack.neutron.v2.domain.FloatingIP> floatingIPOptional = tryFindExistingFloatingIp(neutronFloatingApi, availabilityZone);
+               org.jclouds.openstack.neutron.v2.domain.FloatingIP floatingIP;
+               if (floatingIPOptional.isPresent()) {
+                  floatingIP = floatingIPOptional.get();
+               } else {
+                  floatingIP = createFloatingIpUsingNeutron(neutronFloatingApi, node, availabilityZone);
+               }
+
+               org.jclouds.openstack.neutron.v2.domain.FloatingIP ip = neutronFloatingApi.update(floatingIP.getId(),
+                       org.jclouds.openstack.neutron.v2.domain.FloatingIP.UpdateFloatingIP
+                               .updateBuilder()
+                               .portId(optionalPort.get().getId())
+                               .build());
+
+               input.get().getNodeMetadata().set(NodeMetadataBuilder.fromNodeMetadata(node).publicAddresses(ImmutableSet.of(ip.getFloatingIpAddress())).build());
+            }
```

Comment:
If the node has no ports associated (although unlikely) we'd better at least log a warning instead of silently doing nothing.

Reply:
very good point, shouldn't we fail instead as the user asked to attach a floating IP but jclouds can't satisfy the request because the port is not there?

Reply:
Makes sense too :)